source/vir/src/modes.rs

```diff
@@ -68,90 +68,149 @@ impl Ghost {
     }
 }
 
-struct Typing {
+struct Ctxt {
     pub(crate) funs: HashMap<Fun, Function>,
     pub(crate) datatypes: HashMap<Path, Datatype>,
     pub(crate) traits: HashSet<Path>,
+    pub(crate) check_ghost_blocks: bool,
+    pub(crate) fun_mode: Mode,
+}
+
+// Accumulated data recorded during mode checking
+struct Record {
+    pub(crate) erasure_modes: ErasureModes,
+    // Modes of InferSpecForLoopIter
+    infer_spec_for_loop_iter_modes: Option<Vec<(Span, Mode)>>,
+}
+
+// Temporary state pushed and popped during mode checking
+struct State {
     // for each variable: (is_mutable, mode)
     pub(crate) vars: ScopeMap<Ident, (bool, Mode)>,
-    pub(crate) erasure_modes: ErasureModes,
     pub(crate) in_forall_stmt: bool,
-    pub(crate) check_ghost_blocks: bool,
     // Are we in a syntactic ghost block?
     // If not, Ghost::Exec (corresponds to exec mode).
     // If yes (corresponding to proof/spec mode), say whether variables are tracked or not.
     // (note that tracked may be false even in proof mode,
     // and tracked is allowed in spec mode, although that would be needlessly constraining)
     pub(crate) block_ghostness: Ghost,
-    pub(crate) fun_mode: Mode,
     pub(crate) ret_mode: Option<Mode>,
     pub(crate) atomic_insts: Option<AtomicInstCollector>,
-    // Modes of InferSpecForLoopIter
-    infer_spec_for_loop_iter_modes: Option<Vec<(Span, Mode)>>,
 }
 
-// If we need to catch an error and continue,
-// keep a snapshot of the old transient Typing state fields so we can restore the fields.
-// (used by InferSpecForLoopIter)
-struct TypingSnapshot {
-    vars_scope_count: usize,
-    in_forall_stmt: bool,
-    block_ghostness: Ghost,
-    ret_mode: Option<Mode>,
-    atomic_insts: Option<AtomicInstCollector>,
+struct Typing<'a> {
+    // don't use these fields directly; use * and push_*
+    internal_state: &'a mut State,
+    internal_undo: Option<Box<dyn for<'b> FnOnce(&'b mut State)>>,
 }
```

Comment:
If you place this in a submodule and `use` it at this level, the fields would actually be private to the current module, so there's no risk of someone accidentally accessing `internal_*` directly.